asserts/tool: gpg key pair manager

pedronis opened the pull request:
This starts the work on a helper tool to create assertions, mostly intended to be used by other more specific user tools (snapcraft, model-creator etc).

Introduces a limited key pair manager intended to use/drive a GnuPG setup for the main purpose of signing.

I haven't touched debian/control yet, we need to decide whether this is a separate command or part of snap, and whether it comes with snapd (the package) or a different one. It could even be its own project if we think the overhead in mgmt of that is warranted.

zyga commented on .travis.yml:
```diff
@@ -16,6 +16,7 @@ env:
 before_install:
   - sudo apt-get update -qq
   - sudo apt-get install -qq squashfs-tools
+  - sudo apt-get install -qq gnupg
```
Please add that as a `debian/control` dependency as well

niemeyer commented on asserts/crypto.go:
```diff
@@ -290,14 +303,41 @@ func OpenPGPPrivateKey(privk *packet.PrivateKey) PrivateKey {
+}
+
+// SealedOpenPGPPrivateKey returns a PrivateKey for database use realized as pair of a opengpg packet.PublicKey and a signer function for the sealed private key part.
+func SealedOpenPGPPrivateKey(pubKey *packet.PublicKey, signer func([]byte, *packet.Config) (*packet.Signature, error)) PrivateKey {
```
This feels like a strange interface. Would appreciate a bit more insight into the design here.

niemeyer commented on asserts/export_test.go:
```diff
@@ -28,8 +28,14 @@ import (
 
-// generatePrivateKey exposed for tests
-var GeneratePrivateKeyInTest = generatePrivateKey
+// XXX: just use GenerateKey
```
Yeah, we need to clean this up, but let's talk about that design first so it's more clear how to do that.

niemeyer commented on asserts/tool/gpgkeypairmgr.go:
```diff
@@ -0,0 +1,136 @@
+	cmd.Stderr = &errBuf
+
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("gpg %q failed: %v (%q)", strings.Join(args, " "), err, errBuf.Bytes())
```
Should probably use %s instead of %q as the first one.

niemeyer commented on asserts/tool/gpgkeypairmgr.go:
```diff
@@ -0,0 +1,136 @@
+
+func (gkm *gpgKeypairManager) Put(authorityID string, privKey asserts.PrivateKey) error {
+	// NOTE: we don't need this initially at least and this keypair mgr is not for general arbitrary usage
+	return fmt.Errorf("importing key in a GPG keypair manager is not supported")
```
`"cannot import private key into GPG keyring"`

The original message seems too close to the implementation.

niemeyer commented on asserts/tool/gpgkeypairmgr.go:
```diff
@@ -0,0 +1,136 @@
+		return nil, err
+	}
+	if len(out) == 0 {
+		return nil, fmt.Errorf("no matching key pair found")
```
`"cannot find key %q in GPG keyring"`

niemeyer commented on asserts/tool/gpgkeypairmgr.go:
```diff
@@ -0,0 +1,136 @@
+
+	}
+
+	sign := func(content []byte, cfg *packet.Config) (*packet.Signature, error) {
```
Reading this also gives me a feeling that something might be simplified on this design.

niemeyer commented:
Mostly happy about it, but would like to talk about the design of the sealed private key.

pedronis commented:
I moved GPGKeypairManager to asserts and removed the need of SealedOpenPGPPrivateKey,

also sorted out GenerateKey vs GeneratePrivateKeyInTest (also removed)

pedronis commented:
@niemeyer so I did the simple moving but I thought some more and even if we keep/put GPGKeypairManager in asserts, I want to restructure the internals to look more like the original design I had (but without exporting of course) with clearer boundaries between key pair manager and crypto.go responsibilities:

a) main point: thinking along those lines made it clear that the code needs a bunch more paranoia about the public key properties and the signature properties (same paranoia that we will need on the server when we turn gpg keys into account-keys btw)
b) it will be easier to kick out GPGKeypairManager if we change our mind again
c) from the point of view of setting nice responsibility barriers and right paranoia levels it seems opening simply PrivateKey to external implementation won't never feel quite right, also because we have a bit of a square peg/round hole issue in as much as there will always be things for which low-level RSA signatures are natural and others for which full PGP blobs are natural (tough we may have only one example of the latter in GnuPG itself) so at least in principle we will need adapters/adaption anyway

pedronis commented:
atm this breaks the self-imposed rule that only crypto.go imports openpgp/packet, will work on untangling that without putting the whole key pair mgr in crypto.go

pedronis commented:
reworked, tests (and adding some for the error cases) need a bit more work

mvo5 commented on asserts/gpgkeypairmgr_test.go:
```diff
@@ -0,0 +1,410 @@
+		"timestamp":    time.Now().Format(time.RFC3339),
+	}
+
+	tests := []struct {
```
Nice!

mvo5 commented on asserts/gpgkeypairmgr.go:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	pubKeyBuf := bytes.NewBuffer(out)
+	privKey, err := newExtPGPPrivateKey(pubKeyBuf, gkm.sign)
```
Should we check here that the `keyID` in Get() matches the `privKey.KeyID`, i.e. that gpg gave us what we asked it for? Or is this unneeded?

pedronis replied:
not a failure mode that I would expect (getting 0 or more than one keys are more likely) because we use 0x which forces search by  keyid but we can definitely check in case gpg ships something very buggy

mvo5 commented on asserts/asserts.go:
```diff
@@ -604,7 +604,7 @@ func assembleAndSign(assertType *AssertionType, headers map[string]string, body
 	signature, err := signContent(content, privKey)
 	if err != nil {
-		return nil, fmt.Errorf("failed to sign assertion: %v", err)
+		return nil, fmt.Errorf("cannot sign assertion: %v", err)
```
:heart:

mvo5 commented:
:+1:
